library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java

```diff
@@ -0,0 +1,158 @@
+
+	@Shadow public float zOffset;
+
+	@Inject(method = "renderGuiItemOverlay(Lnet/minecraft/client/font/TextRenderer;Lnet/minecraft/item/ItemStack;IILjava/lang/String;)V",
+			at = @At("HEAD"), cancellable = true)
+	private void quilt$customizeItemOverlay(TextRenderer renderer, ItemStack stack, int x, int y, String countLabel,
+											CallbackInfo ci) {
+		if (stack.isEmpty()) {
+			return;
+		}
+
+		var item = stack.getItem();
+
+		var matrices = quilt$matrices.get();
+		matrices.push();
+		matrices.translate(x, y, 0);
+
+		if (item.preRenderOverlay(matrices, renderer, this.zOffset, stack)) {
+			if (QuiltItemRendering.areOverlayComponentsCustomized(item)) {
+				ci.cancel();
+				renderCustomGuiItemOverlay(matrices, renderer, stack, countLabel);
+			}
+		} else {
+			ci.cancel();
+		}
+
+		matrices.pop();
+	}
+
+	@Redirect(method = "renderGuiItemOverlay(Lnet/minecraft/client/font/TextRenderer;Lnet/minecraft/item/ItemStack;IILjava/lang/String;)V",
+			at = @At(value = "NEW", target = "net/minecraft/client/util/math/MatrixStack"))
+	private MatrixStack quilt$avoidMatrixStackAllocation() {
+		return quilt$matrices.get(); // no need to push now, matrices are only modified when count label is rendered
+	}
+
+	@Redirect(method = "renderGuiItemOverlay(Lnet/minecraft/client/font/TextRenderer;Lnet/minecraft/item/ItemStack;IILjava/lang/String;)V",
+			at = @At(value = "INVOKE", target = "Lnet/minecraft/client/util/math/MatrixStack;translate(DDD)V"))
+	private void quilt$renderCountLabel_pushMatrixStack(MatrixStack instance, double x, double y, double z) {
+		instance.push();
+		instance.translate(x, y, z);
+	}
+
+	@Inject(method = "renderGuiItemOverlay(Lnet/minecraft/client/font/TextRenderer;Lnet/minecraft/item/ItemStack;IILjava/lang/String;)V",
+			at = @At(value = "INVOKE", target = "Lnet/minecraft/client/render/VertexConsumerProvider$Immediate;draw()V",
+					shift = At.Shift.AFTER), locals = LocalCapture.CAPTURE_FAILHARD)
+	private void quilt$renderCountLabel_popMatrixStack(TextRenderer renderer, ItemStack stack, int x, int y, String countLabel,
+													   CallbackInfo ci, MatrixStack matrices) {
+		matrices.pop();
+	}
+
+	@Inject(method = "renderGuiItemOverlay(Lnet/minecraft/client/font/TextRenderer;Lnet/minecraft/item/ItemStack;IILjava/lang/String;)V",
+			at = @At("TAIL"), locals = LocalCapture.CAPTURE_FAILHARD)
+	private void quilt$invokePostRenderOverlay(TextRenderer renderer, ItemStack stack, int x, int y, String countLabel,
+											   CallbackInfo ci, MatrixStack matrices) {
+		if (stack.isEmpty()) {
+			return;
+		}
+
+		matrices.push();
+		matrices.translate(x, y, 0);
+		stack.getItem().postRenderOverlay(matrices, renderer, this.zOffset, stack);
+		matrices.pop();
+	}
+
+	@Unique
+	private void renderCustomGuiItemOverlay(MatrixStack matrices,
+										   TextRenderer renderer, ItemStack stack, @Nullable String countLabel) {
```

Comment:
The line wrapping and formatting on these are a bit weird. I think making your line length longer would fix it. I can't remember if QSL has a limit